`issue edit`, `pr edit`: Support special non-interactive (flags) assignee name `@copilot`

BagToad opened the pull request:
This introduces functionality to handle `@copilot` mentions in assignee flag lists by replacing them with the appropriate login.

BagToad committed:
feat(issue/pr edit): support @copilot in assignee flags

- Introduced CopilotReplacer to handle `@copilot` mentions in assignee lists.

babakks commented on pkg/cmd/pr/shared/editable.go:
```diff
@@ -118,7 +118,28 @@ func (e Editable) AssigneeIds(client *api.Client, repo ghrepo.Interface) (*[]str
-		s := set.NewStringSet()
+		copilotReplacer := NewCopilotReplacer()
+
+		// A closure to replace special assignee names with the actual logins.
+		replaceSpecialAssigneeNames := func(value []string) ([]string, error) {
```
I think that comment is extra, since it's pretty well named.

BagToad replied:
haha yeah good call

BagToad replied:
Fixed in c5f2d7bde70115a15135578a026b765f5c4c3429

babakks commented on pkg/cmd/pr/shared/params.go:
```diff
@@ -312,3 +312,30 @@ func (r *MeReplacer) ReplaceSlice(handles []string) ([]string, error) {
+	return &CopilotReplacer{}
+}
+
+func (r *CopilotReplacer) replace(handle string) (string, error) {
```
I know this is for consistency with `MeReplacer`, but why we need to return an `error`? It's just a lookup and there's no failure if it's not a match.

Feel free to ignore this if you think it's out of scope.

BagToad replied:
Good call, I don't think we need it.

BagToad replied:
Fixed in 871671be526e7b3e40f90a5a562a5eb75136f0a6

babakks commented on pkg/cmd/pr/shared/params.go:
```diff
@@ -312,3 +312,30 @@ func (r *MeReplacer) ReplaceSlice(handles []string) ([]string, error) {
+}
+
+// Replace replaces usages of `@copilot` in a slice with Copilot's login.
+func (r *CopilotReplacer) ReplaceSlice(handles []string) ([]string, error) {
```
The same question.

BagToad replied:
Fixed in 871671be526e7b3e40f90a5a562a5eb75136f0a6

babakks commented on pkg/cmd/pr/shared/params_test.go:
```diff
@@ -187,6 +187,59 @@ func TestMeReplacer_Replace(t *testing.T) {
+	}{
+		{
+			name: "replaces @copilot with copilot-swe-agent",
+			args: args{
```
Since we're dealing with slices, it's good to have two more test cases for `nil` and empty slices.

BagToad replied:
Added in debf0bbaf22b7293da97b10d0ad01f1c64e05904

babakks reviewed (approved):
LGTM! :tada:

andyfeller commented on pkg/cmd/pr/shared/params.go:
```diff
@@ -312,3 +312,30 @@ func (r *MeReplacer) ReplaceSlice(handles []string) ([]string, error) {
+	return &CopilotReplacer{}
+}
+
+func (r *CopilotReplacer) replace(handle string) (string, error) {
+	if strings.EqualFold(handle, "@copilot") {
+		return "copilot-swe-agent", nil
+	}
+	return handle, nil
+}
```
**question:** when should this be reflected in `gh issue` and `gh pr` long description and examples?

BagToad replied:
Excellent catch! Probably in this PR 😄

BagToad replied:
Added in 532aa1c32a8602653f2e12e665e3b50f66d57714

andyfeller reviewed (commented):
Changes seem straight forward.  Aside from other places where `@copilot` changes might apply, I wanted to ask about whether you planned for acceptance testing changes in the final PR.

BagToad committed:
doc(editable): remove needless comment

BagToad committed:
fix(params): remove needless err return

BagToad committed:
test(params): enhance Copilot replacer tests for edge cases

- Added tests for handling nil and empty slices in the Copilot replacer.
- Simplified test structure by removing unnecessary wantErr field.

BagToad committed:
doc(issue/pr edit): doc `@copilot` assignee

andyfeller commented on pkg/cmd/pr/edit/edit.go:
```diff
@@ -60,12 +60,18 @@ func NewCmdEdit(f *cmdutil.Factory, runF func(*EditOptions) error) *cobra.Comman
 			Editing a pull request's projects requires authorization with the %[1]sproject%[1]s scope.
 			To authorize, run %[1]sgh auth refresh -s project%[1]s.
+
+			The %[1]s--add-assignee%[1]s and %[1]s--remove-assignee%[1]s flags both support
+			the following special values:
+			- %[1]s@me%[1]s: assign or unassign yourself
+			- %[1]s@copilot%[1]s: assign or unassign Copilot
```
**question:** should we mention that `@copilot` at this time does not work with `--add-reviewer` or `--remove-reviewer` flags in support of https://github.blog/changelog/2024-10-29-github-copilot-code-review-in-github-com-private-preview/?

andyfeller replied:
I wonder if we should actively error with a specific error message if someone uses `@copilot` in this use case even.

BagToad replied:
> **question:** should we mention that `@copilot` at this time does not work with `--add-reviewer` or `--remove-reviewer` flags in support of https://github.blog/changelog/2024-10-29-github-copilot-code-review-in-github-com-private-preview/?

I don't have a strong opinion on this one - though I loosely think not stating that it is supported might be enough. Again, I don't have a strong opinion, so I'm happy to add it if that is where you are leaning. Let me know 😁 


> I wonder if we should actively error with a specific error message if someone uses `@copilot` in this use case even.

Here's the current error experience:

```
❯ gh pr edit 2 --add-reviewer "@copilot"
'@copilot' not found
```

andyfeller replied:
**suggest:** Middle of the road might be adding a line in the long description, which we can point users to it if they try to use the feature and don't understand _"`@copilot` not found"_ really means _"You cannot use `@copilot` as a reviewer currently"_

BagToad replied:
What do you think about this:

```suggestion
			The %[1]s--add-assignee%[1]s and %[1]s--remove-assignee%[1]s flags both support
			the following special values:
			- %[1]s@me%[1]s: assign or unassign yourself
			- %[1]s@copilot%[1]s: assign or unassign Copilot
			
			The %[1]s--add-reviewer%[1]s and %[1]s--remove-reviewer%[1]s flags do not support
			these special values.
```

BagToad replied:
What do you think about this:

```suggestion
			The %[1]s--add-assignee%[1]s and %[1]s--remove-assignee%[1]s flags both support
			the following special values:
			- %[1]s@me%[1]s: assign or unassign yourself
			- %[1]s@copilot%[1]s: assign or unassign Copilot
			
			The %[1]s--add-reviewer%[1]s and %[1]s--remove-reviewer%[1]s flags do not support
			these special values.
```

Since I realized `@me` is also not supported.

```
gh pr edit 3 --add-reviewer @me
'@me' not found
```

BagToad replied:
Updated in 748afb6e85cf56bd6ad4ae6f348dcfd7d105192f

BagToad committed:
doc(issue/pr edit): clarify @copilot usage

andyfeller reviewed (approved):
![shipit-squirrel-stareyes](https://github.com/user-attachments/assets/b65ac61f-6ab4-4caf-910e-15cd2a0061c6)

snstgithub reviewed (commented):
- [ ]
